applications/OptimizationApplication/python_scripts/responses/binary_operator_response_function.py

```diff
@@ -0,0 +1,96 @@
+from KratosMultiphysics.OptimizationApplication.utilities.response_utilities import EvaluateValue
+from KratosMultiphysics.OptimizationApplication.utilities.response_utilities import EvaluateGradient
+from KratosMultiphysics.OptimizationApplication.utilities.response_utilities import BinaryOperator
+
+class BinaryOperatorResponseFunction(ResponseFunction):
+    def __init__(self, response_function_1: ResponseFunction, response_function_2: ResponseFunction, binary_operator: BinaryOperator, optimization_problem: OptimizationProblem):
+        if binary_operator == BinaryOperator.DIVIDE:
+            # this is because, the optimization_problem data container uses "/" as a path separator.
+            super().__init__(f"({response_function_1.GetName()}÷{response_function_2.GetName()})")
+        else:
+            super().__init__(f"({response_function_1.GetName()}{binary_operator.value}{response_function_2.GetName()})")
+
+        self.optimization_problem = optimization_problem
+        self.response_function_1 = response_function_1
+        self.response_function_2 = response_function_2
+        self.binary_operator = binary_operator
+        self.model_part: Optional[Kratos.ModelPart] = None
+
+    def GetImplementedPhysicalKratosVariables(self) -> 'list[SupportedSensitivityFieldVariableTypes]':
+        vars_list = self.response_function_1.GetImplementedPhysicalKratosVariables()
+        vars_list.extend(self.response_function_2.GetImplementedPhysicalKratosVariables())
+        return vars_list
+
+    def Initialize(self) -> None:
+        self.response_function_1.Initialize()
+        self.response_function_2.Initialize()
+
+        if len(self.response_function_1.GetImplementedPhysicalKratosVariables()) != 0 and len(self.response_function_2.GetImplementedPhysicalKratosVariables()) != 0:
+            self.model_part = ModelPartOperation(self.response_function_1.GetInfluencingModelPart().GetModel(), ModelPartOperation.OperationType.UNION, f"response_{self.GetName()}", [self.response_function_1.GetInfluencingModelPart().FullName(), self.response_function_2.GetInfluencingModelPart().FullName()], False).GetModelPart()
+        elif len(self.response_function_1.GetImplementedPhysicalKratosVariables()) != 0:
+            self.model_part = self.response_function_1.GetInfluencingModelPart()
+        elif len(self.response_function_2.GetImplementedPhysicalKratosVariables()) != 0:
+            self.model_part = self.response_function_2.GetInfluencingModelPart()
+
+    def Check(self) -> None:
+        self.response_function_1.Check()
+        self.response_function_2.Check()
+
+    def Finalize(self) -> None:
+        self.response_function_1.Finalize()
+        self.response_function_2.Finalize()
+
+    def GetInfluencingModelPart(self) -> Kratos.ModelPart:
+        return self.model_part
+
+    def CalculateValue(self) -> float:
+        v1 = EvaluateValue(self.response_function_1, self.optimization_problem)
```

Comment:
I don't understand the concept of evaluate. Why can't we just use Response Routine and it deifnes if smth should be computed or return from value directly ? It looks like dublication or bad design of response / response routine .

Reply:
Assume the scenario where "r1 + r2 * r1", where r1 needs to be evaluated twice. Instead of evaluating them twice, I use the EvaluteValue method to store the first evaluation of r1, and then call it in the second evaluation. I agree to your point that, instead of saving it in the OptProb, under respones, I will use a different place to save it. So, saving the value under the OptProb is decided by the response routine, and the call to CalculateValue will be normal per one evaluation.

Reply:
Same for the gradients

Reply:
I would suggest using the Response Routine in this class, so then you can solve r1 (x_i) + r2 (x_i-1)) overwise, we are fixed with only current x_i ... And the Response Routine already has all the functionality to check if the model has to be run or it is already solved and we can get value immediately.